### ISSUE TITLE
Clenshaw's algorithm (documentation, optimization)

### ISSUE THREAD

**Issue by jrus** [chebfun/chebfun#1582]
Looking at the guts of Chebfun’s implementation, it’s clear that the main way chebtechs are evaluated is via Clenshaw’s algorithm, but in the documentation, website, publications, etc. there’s hardly any mention of it. In the ATAP book, it’s relegated to one brief exercise at the end of chapter 3. By contrast, there’s extensive discussion of the barycentric interpolation formula all over the place. This might give the impression to someone who doesn’t dive into the code that the barycentric formula is used for routine evaluation.

It would be nice somewhere to see a discussion of the relative speed, accuracy, etc. of using the barycentric formula on chebyshev points vs. converting from points -> coefficients and then using Clenshaw’s algorithm instead. In what practical cases does it make sense to use the barycentric formula for evaluations? (Seems to me that doing all those divisions is pretty expensive if they can be avoided.) Etc.

Semi-related: has optimizing Chebfun’s implementation of Clenshaw’s algorithm been investigated? I’m not an expert in what trickery Matlab will do behind the scenes for straight-forward looking code, so maybe this is already getting optimized by Matlab, but this seems like the kind of workload that can be easily parallelized to run on multiple CPU cores or a GPU, or at least JIT compiled into an efficient little kernel. Would it be worth doing some profiling and experimentation, or has there already been work on this? (I do notice the loop has been unrolled to do two steps at each iteration.)

Also semi-related, in sleuthing around Google scholar I came across this paper about a scheme for doing precise evaluations using Clenshaw’s algorithm which avoids some roundoff errors (especially near roots) at the expense of a big pile of extra multiply and add operations. You folks have probably seen it, but I thought I’d throw in a link:
http://www.sciencedirect.com/science/article/pii/S0096300311006242
http://www.researchgate.net/profile/Roberto_Barrio/publication/220562750_Accurate_evaluation_of_a_polynomial_in_Chebyshev_form/links/53e322eb0cf2b9d0d832cdbe.pdf

**Comment by trefethen** [comment]
Jacob, you raise some interesting issues; let me comment about the history.  Chebfun was born from the realization that almost all the world thought polynomial interpolation was numerically impossible, yet in fact it's as stable an algorithm as any on earth if you use Chebyshev points and the barycentric formula.  In the early years our focus was very much on the barycentric formula.

As time has passed, the world has to some extent gotten the message that polynomials are ok, and the Chebfun team has totally gotten the message and developed a lot of experience with such matters.  As the experience has developed, we've found that in virtually every case, you can do things "by values" (e.g. barycentric interpolation formula) or "by coefficients" (e.g. Clenshaw's algorithm).  As time has passed we've deemphasized values and used coefficients in more places, without having exceptionally strong feelings in either direction; nor has this seemed an exciting enough transition for us to focus on the matter in our research articles.  So you're right, today's actual Chebfun code, which uses coefficients a lot, doesn't match very well the published material, which speaks of values a lot.

Maybe the place where the values/coefficients discussion is most alive these days would be in the solution of differential equations by "collocation" or "ultraspherical" methods.  Here too it seems that many things can be done approximately equally well with either technology.

**Comment by jrus** [comment]
Thanks for the clarification.

> As time has passed we've deemphasized values and used coefficients in more places, without having exceptionally strong feelings in either direction

This is roughly what I’d gathered from looking around in the code and reading your books/papers.

So is it fair to say that in general working in terms of either values or coefficients will give about equally precise and numerically stable answers, and the choice comes down to “do whichever is more convenient” or maybe “time it and see which one is faster for a particular use case”?

**Comment by jrus** [comment]
It’s possible to make some slight improvement to Chebfun’s clenshaw implementation by putting the test for odd degree polynomials up front instead of at the end, and for odd polynomials just setting `b_{n+1} = c_n` instead of `0`. This is also enough of a hot code path that I don’t think it‘s worth splitting the scalar and vector cases behind function calls.

Here’s my change:
```matlab
function y = clenshaw(x, c)
%CLENSHAW   Clenshaw's algorithm for evaluating a Chebyshev expansion. [...]
if ( ~isnumeric(x) || ~isnumeric(c) )
    error('CHEBFUN:CHEBTECH:clenshaw:inputs', ...
        ['Evaluation points should be a column vector, and coefficients ' ...
         'should be a matrix.']);        
elseif ( ~iscolumn(x) )
    warning('CHEBFUN:CHEBTECH:clenshaw:xDim', ...
        'Evaluation points should be a column vector.');
    x = x(:);
end

% Clenshaw scheme for scalar-valued functions.
if ( iscolumn(c) )
    n = numel(c);
    x = 2*x;
    b2 = zeros(numel(x), 1);
    b1 = b2;
    if ( ~mod(n, 2) )
        b1(:) = c(n);  % For odd-degree polynomials, handle last term separately.
        n = n - 1;
    end
    for k = n:-2:3
        b2 = c(k) + x.*b1 - b2;
        b1 = c(k-1) + x.*b2 - b1;
    end
    y = c(1) + .5*x.*b1 - b2;

% Clenshaw scheme for array-valued functions.
elseif ( ismatrix(c) )
    [n, cColumns] = size(c);
    x = 2*x;
    b2 = zeros(numel(x), cColumns);
    b1 = b2;
    if ( ~mod(n, 2) )
        b1 = repmat(c(n,:), numel(x), 1);
        n = n - 1;
    end
    for k = n:-2:3
        b2 = bsxfun(@plus, c(k,:), bsxfun(@times, x, b1) - b2);
        b1 = bsxfun(@plus, c(k-1,:), bsxfun(@times, x, b2) - b1);
    end
    y = bsxfun(@plus, c(1,:), .5 * bsxfun(@times, x, b1) - b2);

else
    error('CHEBFUN:CHEBTECH:clenshaw:cDim', ...
        'Coefficients must be a column vector or matrix.');
end

end
```

Not sure this is the right context for discussing this, but anyway.. :)

**Comment by nickhale** [comment]
With the tests in [this gist](https://gist.github.com/nickhale/455cb3c9ef6a2ba1b3a6) I get the following:

When `f = chebtech2(@(x) cos(1000*pi*x)+sin(1000*pi*x));`
```
test1
Elapsed time is 0.026949 seconds. % <-- f.clenshaw
Elapsed time is 0.028254 seconds. % <-- clenshaw_new (from above)
test2
Elapsed time is 0.025296 seconds.
Elapsed time is 0.024570 seconds.
test3
Elapsed time is 0.338482 seconds.
Elapsed time is 0.365244 seconds.
test4
Elapsed time is 0.511062 seconds.
Elapsed time is 1.049839 seconds.
```

When `f = chebtech2(@(x) cos(pi*x) + sin(pi*x));`
```
test1
Elapsed time is 0.005536 seconds.
Elapsed time is 0.003798 seconds.
test2
Elapsed time is 0.001238 seconds.
Elapsed time is 0.001216 seconds.
test3
Elapsed time is 0.014126 seconds.
Elapsed time is 0.010211 seconds.
test4
Elapsed time is 0.005980 seconds.
Elapsed time is 0.008738 seconds.
```

In summary, it seems that the proposed change is in general a little slower for long representations and a little faster for short ones. The exception is in the case of array-valued representations, where the proposed code is a factor of two slower. I suspect the main difference is in the use of `bsxfun` rather than `repmat`. Some hybrid of the two implementations might turn out the be faster. 

I should also note that the proposed code was **very** slow when evaluating at a single point. (Not sure why.) In any case, I think both the new and old codes can be improved by treating the scalar `x` matrix `c` inputs as a special case. (In fact, the results I show above have this implemented.)

```
if ( iscolumn(c) )
    ...

elseif ( isscalar(x) )
    [n, cColumns] = size(c);
    x = 2*x;
    b1 = zeros(1, cColumns);
    b2 = zeros(1, cColumns);
    if ( ~mod(n, 2) )
        b1(:) = c(n,:);  % For odd-degree polynomials, handle last term separately.
        n = n - 1;
    end
    for k = n:-2:3
        b2 = c(k,:) + x*b1 - b2;
        b1 = c(k-1,:) + x*b2 - b1;
    end
    y = c(1,:) + .5*x.*b1 - b2;    

% Clenshaw scheme for array-valued functions.
elseif ( ismatrix(c) )
 ...
```

**Comment by jrus** [comment]
I guess I didn’t test with too broad a range of inputs.

The first thing to do is probably make a set of benchmarks with a large variety of different inputs, look for which ones seem to be unusually slow in flop count per second (or whatever metric), and then profile those.

Here’s a partial list to start:

- evaluation of a small 1-piece chebfun at a large number of points
- evaluation of a large 1-piece chebfun at a large number of points
- many repeated evaluations of a small chebfun at a few points in a loop
- many repeated evaluations of a large chebfun at a few points

- same as the above conditions but with a non-standard domain (not [-1,1])

- same as the above conditions but different types of funs (singfuns, trigtechs, ...)

- try all of the above with both real and complex evaluation point vector inputs

- same as all the above but with imaginary-valued or complex-valued chebfuns

- evaluation of a chebfun with two pieces at a large number of points
- many repeated evaluations of a chebfun with two pieces at a few points
- evaluation of a chebfun with many small pieces at a large number of points
- many repeated evaluations of a chebfun with many small pieces at a few points

**Comment by jrus** [comment]
It’s also worth benchmarking the scalar loop version, something like:
```matlab
clen = numel(c);
xlen = numel(x);
y = zeros(xlen);
for xIdx=1:xlen
    xi = x[xIdx];
    xi2 = 2*xi;
    b2 = 0;
    b1 = 0;
    n = clen;
    if ( ~mod(n, 2) )
        b1 = c(n);
        n = n - 1;
    end
    for k = n:-2:3
        b2 = c(k) + xi2*b1 - b2;
        b1 = c(k-1) + xi2*b2 - b1;
    end
    y[xIdx] = c(1) + xi*b1 - b2;
end
```
Which may end up being faster for certain input sizes.

**Comment by jrus** [comment]
I asked my wife, who is a C programmer, to help with a C implementation, so we have (mostly she has) been working on that a bit today. Here is a C version which uses one [SIMD](https://en.wikipedia.org/wiki/SIMD) [FMA](https://en.wikipedia.org/wiki/FMA_instruction_set) instruction plus one SIMD addition per coefficient per evaluation point: [github link][https://github.com/thinkingfish/numcrunch/blob/master/src/clenshaw_vector.c]. It runs about 40% faster on my laptop than the Chebfun Matlab version (for large vectors of both coefficients and evaluation points), when the C version is running on one core, and Matlab is run in parallel across 4 logical cores. Sometime soon we’ll try to parallelize this code, and maybe try to do a bit of chunking of coefficients to do a better job keeping them in CPU cache for longer, which I suspect can provide about 2x speedup for large coefficient vectors. I’m guessing (but so far without direct proof) that we’ll be able to make a C version which runs on the CPU ultimately about 10x faster than the Matlab version for large vectors, and considerably better than that for small vectors, because Matlab takes quite a bit more fixed overhead up front.

**Comment by jrus** [comment]
Oh excellent: someone has written a little benchmark harness already: http://www.mathworks.com/matlabcentral/fileexchange/18798-timeit-benchmarking-function

**Comment by jrus** [comment]
I’ve been playing around with timing various stuff, and I must say, Matlab makes certain operations weirdly slow. For example saying `b1 = 0*x` is apparently much faster than saying `b1 = zeros(numel(x), 1)` or saying `b1(numel(x)) = 0`. Anyway, I have a lot more timing to do, but here’s my latest version:

```matlab
function y = clenshaw_jrus(x, c)

n = size(c, 1);
x = 2*x;
b1 = 0*x;
b2 = b1;
if ( ~mod(n, 2) )
    b1 = b1 + c(n);  % For odd-degree polynomials, handle last term separately.
    n = n - 1;
end
for k = n:-2:3
    b2 = c(k) + x.*b1 - b2;
    b1 = c(k-1) + x.*b2 - b1;
end
y = c(1) + .5*x.*b1 - b2;

end
```

**Comment by jrus** [comment]
Also, Yao added multithreading to her C version, which seems to scale almost linearly with the number of cores. On her 2013 Apple MBP laptop (2.8 GHz Intel Core i7), that means about 4x speedup over the single threaded version. See the latest version of her readme at https://github.com/thinkingfish/numcrunch/

So now that’s about 5–6 times faster than Matlab overall.

I think we should still be able to push that a bit faster via some chunking of the coefficient vector to better keep the computation in the CPU cache, but we haven’t tested, so it’s hard to tell exactly how much it will help.

Figuring out how to make an OpenCL or other GPU version is going to take more work, as neither of us is an expert in GPU programming, but this is a straight-forward enough problem that it should make a good learning exercise.

**Comment by nickhale** [comment]
With the following I find the old and the new codes to be comparable on my machine, with a slight win for @jrus's new code. Since the new code is also cleaner (IMO) I'd be in favour of switching to it.

https://gist.github.com/nickhale/c26fc355294c52112da3

**Comment by jrus** [comment]
I’m still going to do some experimenting today about the array-valued chebfun version.

Here’s my code for making a plot over a bunch of coefficient sizes:
https://gist.github.com/jrus/76cbb5295a16f33abd64

**Comment by jrus** [comment]
One more slight change, that noticeably helps in the linear (degree 1) case:
```matlab
n = size(c, 1);
x = 2*x;
b2 = 0*x;
if ( ~mod(n, 2) )
    b1 = b2 + c(n);
    n = n - 1;
else
    b1 = b2;
end
for k = n:-2:3
    b2 = c(k) + x.*b1 - b2;
    b1 = c(k-1) + x.*b2 - b1;
end
y = c(1) + .5*x.*b1 - b2;

end
```

**Comment by jrus** [comment]
The scalar loop version turns out to be faster on my machine for numbers of evaluation points `x` between about 2–35. For evaluating at 4 `x` points for instance, there’s a very dramatic 8x speed difference:
<img width="672" alt="screen shot 2015-10-05 at 3 16 23 pm" src="https://cloud.githubusercontent.com/assets/83411/10295247/38fc2746-6b74-11e5-9b31-0abd5e65d389.png">
For 1 points (where both are treated as scalar versions), or for 36 points, the two versions run at about the same speed. For >36 points, the vector version has the advantage.

```matlab
function y = clenshaw_scalar_loop(x, c)

n = numel(c);
b1init = 0;
if ( ~mod(n, 2) )
    b1init = b1init + c(n);
    n = n - 1;
end
for xidx=numel(x):-1:1
    xi = x(xidx);
    xi2 = 2*xi;
    b2 = 0;
    b1 = b1init;
    for k = n:-2:3
        b2 = c(k) + xi2*b1 - b2;
        b1 = c(k-1) + xi2*b2 - b1;
    end
    y(xidx) = c(1) + xi*b1 - b2;
end

end
```

**Comment by jrus** [comment]
Oh, I didn’t realize @nickhale wrote a C implementation of part of Chebfun over at <https://github.com/nickhale/libchebfun/>.

Here’s the relevant vector clenshaw implementation,
[libchebfun/src/util.c#L340-L393](https://github.com/nickhale/libchebfun/blob/master/src/util.c#L340-L393) vs. Yao’s [numcrunch/src/clenshaw_vector.c#L40-L73](https://github.com/thinkingfish/numcrunch/blob/master/src/clenshaw_vector.c#L40-L73).

But it’s now a few years out of date, so it doesn’t use fused multiply–subtract operations or larger vector sizes available with current Intel consumer CPUs, or any multithreading, which should make a substantial difference in running speed. My suspicion is that that C version will run slightly but not overwhelmingly faster than the current Matlab for large problems or even a bit slower on machines with several cores (but with much less overhead than Matlab for small problems).

A production-ready version will probably need to check for various CPU capabilities in a compile-time check.

**Comment by jrus** [comment]
It’s probably not often practically useful, but if there’s ever a need to evaluate a very very high degree polynomial (degree in the millions or more, high enough that it becomes a performance bottleneck) at a small number of points, then an implementation which evaluates a polynomial by handling blocks of the coefficients in parallel could be handy. Here’s one paper on the subject, http://epubs.siam.org/doi/pdf/10.1137/S1064827596312857

There don’t seem to be many others on precisely the subject, though of course there are many papers about solving tridiagonal systems or evaluating general recurrences on various kinds of parallel architectures.

**Comment by jrus** [comment]
I’ve been working on getting a C version working from inside Matlab. Since I’m not a C programmer, it took some help and some teeth gnashing, but I managed to get a basic version working for real inputs, and only running on one CPU core. As it is, it runs about 3x faster than the Matlab version, but that should scale almost linearly with logical cores, so on my laptop that would be 12x faster, and on a desktop machine it might be 24x faster, or more. Moreover, there’s still some performance improvement to eke out by making better use of the CPU cache hierarchy, I think potentially 1.5–2x.

Compiling on my Mac is slightly tricky, because apparently the new versions of OS X and XCode are not recognized within the XML configuration file Matlab uses to pick a compiler. The 2016 version of Matlab should fix that, but in the mean time [this Matlab support ticket](http://www.mathworks.com/matlabcentral/answers/246507-why-can-t-mex-find-a-supported-compiler-in-matlab-r2015b-after-i-upgraded-to-xcode-7) has a downloadable replacement XML file. That still doesn’t quite cut it, because by default it adds `-arch x86_64` as a compiler flag, which seems to disable more recent SIMD intrinsics. Putting in `-march=native` instead apparently fixes that.

Anyway, I’ll try to make a branch that folks can play with sometime soon. I was going to attach the compiled Mac binary here, but github doesn’t support attaching such files.

**Comment by jrus** [comment]
Okay, I have a C version which as far as I can tell entirely compatible with the Matlab version (modulo some tiny rounding differences from using FMA instructions, and some better handling of a few edge cases), on a branch:

https://github.com/jrus/chebfun/blob/feature-clenshaw-c-simd/%40chebtech/clenshaw.c
https://github.com/jrus/chebfun/blob/feature-clenshaw-c-simd/%40chebtech/clenshaw.h

Here’s a compiled Mac binary, which works on OS X 10.9.5 on an MacBookPro11,1 laptop with Intel Core i5 CPU:
https://github.com/jrus/chebfun/blob/feature-clenshaw-c-simd/%40chebtech/clenshaw.mexmaci64

With that in place, all Chebfun tests pass, and there's at least a bit of speedup across all tests.

This is definitely not the most heavily optimized possible C code: it’s just running on a single core, and it doesn’t try to be clever about blocking the work to stay in cache. But it’s a reasonable first baseline. It’s also just slightly tricky to get Matlab to compile (see the previous comment), but that can also be worked on.

Anyone with a Mac, try dropping that binary in your `chebfun/@chebtech/` directory and see if it works for you too. Or if anyone has a Windows machine and wants to try to get it compiling in MSVC, that could also be fun.

**Comment by jrus** [comment]
I added support for 256-bit SIMD vectors without FMA support (AVX), 128-bit SIMD vectors (SSE2), and a plain scalar version, via some ifdefs in the header file.

@nickhale figured out yesterday that trying to run a compiled binary which expects FMA support on a 2011-era CPU causes a segfault (whoops!) I emailed him and @aaustin141 a compiled binary of the AVX-but-no-FMA version, and would be happy to email various compiled versions of this code to anyone who wants to try it. [I don’t have a Windows machine at the moment.]

I think it should be relatively straightforward to compile about 3 versions for each of {OS X, Windows, Linux}, and then make a little Matlab script which checks for available CPU feature support, and copies then appropriate binary into place. I think it should be possible to distribute binaries, and not require any explicit compilation from end users, but I’m far from a MEX or C expert, so we’ll see.

**Comment by jrus** [comment]
Here’s an example plot from my laptop showing what kind of speed we get when evaluating at four points `X`:
<img width="617" alt="screen shot 2015-11-12 at 9 51 50 pm" src="https://cloud.githubusercontent.com/assets/83411/11139717/b93b241c-8987-11e5-858c-882be6f3ef61.png">

Black is the current implementation; green is my proposal from upthread where the even/odd check makes the highest-degree odd coefficient get treated separately, instead of doing the conditional at the end; red is a scalar loop [remember, this is fast for small numbers of `X` points, but becomes worse for large numbers of points]; and blue is the compiled C version (with 256-bit vectors and FMA support). I stripped out any error handling or other code from the Matlab versions and I copied all these functions into their own folder instead of running anything through Chebfun, to better compare just their numeric performance. (Though the compiled version still includes a bunch of error handling which probably costs a bit of overhead for small problems.)

The vertical axis represents the log base two of number of evaluation points times number of coefficients divided by runtime per call, i.e. `log2(xlen * clen / time)`. We can see that for 4 evaluation points, after we get past the initial overhead, the scalar loop version is about 8 times as fast as the current version, and the compiled C version is about 15 times faster than that.

I’m actually slightly surprised that there isn’t a bigger difference in fixed overhead between the Matlab and C versions. For very tiny problems, the compiled version is only about twice as fast as the Matlab version. The Matlab version gets down to 7.7 microseconds to compute the output when there is only one coefficient, whereas the C version is 3.1 microseconds. I should maybe try testing with all the error handling included.

* * *

If we use 64 evaluation points instead of 4, we can see that the scalar loop version falls behind, and the compiled version is only about 9x faster than the current version:

<img width="621" alt="screen shot 2015-11-12 at 10 25 55 pm" src="https://cloud.githubusercontent.com/assets/83411/11140178/700098b8-898c-11e5-87f3-e5609374d5de.png">

Now the compiled version is only about 9.7x faster than the current version for problems with 2^16 coefficients, whereas with 4 points it was on order 100x faster.

* * *

If we go up to 512 evaluation points, the scalar loop version falls hopelessly behind.

The compiled version is a bit slow for problems with a small even number of coefficients. I’m not quite sure what’s happening there, but I’m sure that can be fixed with a bit of optimization effort. For high-degree polynomials it ends up only about 2.5x faster. (Remember, again, this is running on only one core, whereas I believe Matlab is running across all 4 logical cores on my machine. Also the code is pretty naïve; I suspect it can be made considerably faster with some careful cache management.)

<img width="588" alt="screen shot 2015-11-12 at 10 46 48 pm" src="https://cloud.githubusercontent.com/assets/83411/11140457/848a76c0-898f-11e5-90e8-1713a000b497.png">
